### ISSUE TITLE
[Libtorch of C/C++ version] 'to_dense()' function have a Bug. in C/C++

### ISSUE THREAD

**Issue by TaeSuMoon** [pytorch/pytorch#104021]
### 🐛 Describe the bug

# this code used python and it's very well work

          import torch
          
          intrinsic = torch.tensor([3,4,5,6]).cuda()
          idx = torch.tensor([[0, 0, 1, 1], [0, 2, 1, 2]]).cuda()
          
          s = torch.sparse_coo_tensor(idx, intrinsic, (2, 3))
          print(s)
          print("")
          print(s.to_dense())
          print("")
          
          K = torch.cat([s.to_dense(), torch.tensor([[0, 0, 1]]).cuda()], dim=0)
          print(K)

# But, i used C/C++ code in visual studio, as a result of have a BUG!
# my versions : VS2017(v141) // compiler is (/std:c++14) // libtorch ver.1.9.0 // CUDA 10.2
# this code is C/C++

          #include <cuda_runtime.h>
          #include <device_launch_parameters.h>
          #include <cuda.h>
          #include <Eigen/Dense>
          #include <torch/torch.h>
          
          torch::Tensor idx = torch::tensor({ { 0,0,1,1 }, { 0,2,1,2 } }).to(torch::kCUDA);
          torch::Tensor local_intrinsic = torch::tensor({3,4,5,6}).to(torch::kCUDA);
          torch::IntArrayRef size = { 2, 3 };
          torch::Tensor s = torch::sparse_coo_tensor(idx, local_intrinsic, size).to(torch::kCUDA);
          cout << s << endl;
          torch::Tensor K = s.to_dense().to(torch::kCUDA);

# line for 'torch::sparse_coo_tensor()' is very well work but, next line have a bug
# occur bug line is '.to_dense()'
# i don't know WHY NOT WORK THIS FUNCTION.
# AND Error Message is ==> 'vector subscript out of range',  BUT C++ CODE IS NOT WRONG.

Thank u :)

### Versions

#written above#

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jbschlosser

**Comment by amjames** [comment]
@TaeSuMoon when you run the python sample that does not produce an error is that also using torch version 1.9.0? 

That version is quite out of date so it is possible that this bug has been resolved and updating libtorch will resolve your issue. I can not reproduce your issue with a current build.

**Comment by alexsamardzic** [comment]
Try, for example, following:

```C++
#include <torch/torch.h>

#include <iostream>
#include <vector>

int main() {
    torch::Tensor idx = torch::tensor({ { 0,0,1,1 }, { 0,2,1,2 } }).to(torch::kCUDA);
    torch::Tensor local_intrinsic = torch::tensor({3,4,5,6}).to(torch::kCUDA);
    std::vector<int64_t> vec_size = { 2, 3 };
    torch::IntArrayRef size{ vec_size };
    torch::Tensor s = torch::sparse_coo_tensor(idx, local_intrinsic, size).to(torch::kCUDA);
    std::cout << "sparse = " << s << std::endl;
    torch::Tensor K = s.to_dense();
    std::cout << "dense = " << K << std::endl;

    return 0;
}
```

The problem in your code is with `size` variable initialization, for various options you could check `ArrayRef` constructors, in `c10/util/ArrayRef.h` file of PyTorch source tree.
